File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

int solution(string s) {
    int answer = 0;
    
    int n=s.size();
    s+=s;
    
    for(int i=0;i<n;i++){
        string str="";
        for(int j=i;j<i+n;j++){
            str+=s[j];
        }

        
        stack<char>st;
        for(auto a:str){
            
            if(st.empty()){
                st.push(a);
                continue;
            }
            
            char ch=st.top();
            
            if(ch=='{'&&a=='}'){
                st.pop();
            }else if(ch=='['&&a==']'){
                st.pop();
            }else if(ch=='('&&a==')'){
                st.pop();
            }
            else{
                st.push(a);
            }
        }
        
        if(st.empty()){
            answer++;
        }
        
    }
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    
    public int solution(String s) {
        
        int answer = 0;
        
        for(int i=0;i<s.length();i++){
            String str = s.substring(i,s.length())+s.substring(0,i);
            
            if(isValid(str))
                answer++;
            // System.out.println(str+" "+answer);
        }
        
        return answer;
    }
    
    private boolean isValid(String str){
    
        Stack<Character> st = new Stack<>();
        
        for(char ch : str.toCharArray()){
                        
            if(st.isEmpty()){
                st.add(ch);
                continue;       
            }
            
            char peek = st.peek();
            
            switch(peek){
                case '[':
                    if(ch==']'){
                        st.pop();
                        break;
                    }   
                case '(':
                    if(ch==')'){
                        st.pop();
                        break;
                    }
                case '{':
                    if(ch=='}'){
                        st.pop();
                        break;
                    }
                default:
                    st.add(ch);
            }
        }
        
        return st.isEmpty();
    }
    
    
}